rasterio/windows.py

```diff
@@ -592,28 +592,32 @@ def toslices(self, shape=None):
 
         Parameters
         ----------
-        arr: ndarray
+        shape: tuple
+            The number of rows and cols of an array to be sliced.
 
         Returns
         -------
         row_slice, col_slice: slice
             A pair of slices in row, column order
 
         """
-        if shape is None:
-            return tuple(slice(*rng) for rng in self.toranges())
-        else:
+        (r0, r1), (c0, c1) = self.toranges()
+
+        if shape is not None:
             height, width = shape
-            return (
-                slice(
-                    int(math.floor(self.row_off - height)),
-                    int(math.ceil(self.row_off - height + self.height)),
-                ),
-                slice(
-                    int(math.floor(self.col_off - width)),
-                    int(math.ceil(self.col_off - width + self.width)),
-                ),
-            )
+            if r0 < 0:
+                r0 = r0 - height
```

Comment:
Previously in this case, we didn't switch to end-indexing and the slices were broken. Always subtracting the array height would have worked too, but user code might be thrown off by unexpected (but correct) negative values.